tools/crm_rule.c

```diff
@@ -63,16 +63,19 @@ crm_rule_check(pe_working_set_t *data_set, const char *rule_id, crm_time_t *effe
     /* Rules are under the constraints node in the XML, so first find that. */
     cib_constraints = get_object_root(XML_CIB_TAG_CONSTRAINTS, data_set->input);
 
-    /* Get all rules matching the given ID, which also have only a single date_expression
-     * child whose operation is not 'date_spec'.  This is fairly limited, but it's hard
-     * to check expressions more complicated than that.
+    /* Get all rules matching the given ID which are also simple enough for us to check.
+     * For the moment, these rules must only have a single date_expression child and:
+     * - Do not have a date_spec operation, or
+     * - Have a date_spec operation that contains years= but does not contain moon=.
      */
-    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec']", rule_id);
+    xpath = crm_strdup_printf("//rule[@id='%s']/date_expression[@operation!='date_spec'] |"
+                              "//rule[@id='%s']/date_expression[@operation='date_spec' and date_spec/@years and not(date_spec/@moon)]",
+                              rule_id, rule_id);
     xpathObj = xpath_search(cib_constraints, xpath);
 
     max = numXpathResults(xpathObj);
     if (max == 0) {
-        CMD_ERR("No rule found with ID=%s containing a date_expression", rule_id);
+        CMD_ERR("No rule found with ID=%s, or rule does not have a date_expression, or rule is too complicated", rule_id);
```

Comment:
I'm thinking a more specific message would be worth the overhead of more xpath searches, which shouldn't really be noticeable since crm_rule will only ever be called manually (vs in a loop of some sort or at a time-sensitive moment). We could even potentially do the extra searches only if the first one doesn't turn up anything.

Reply:
Okay, I can do that.  Would you like more tests for all these possibilities, too?

Reply:
Might as well